kala.py

```diff
@@ -37,43 +39,185 @@
+    return document
+
+
+def _convert_object(document):
+    """This is a wrapper for _convert_object_type()
+
+    :param document:
+    document -- This should either be a JSON document or a list of JSON documents.
+    """
+    document = _convert_object_type(document, 'ISODate')
+    document = _convert_object_type(document, 'UUID')
+    return document
+
+
+@app.route('/aggregate/<collection>', method=['GET'])
+def get_aggregate(mongodb, collection):
+    if bottle.request.method == 'GET':
+        pipeline = _get_json('pipeline')
+        # Should this go in the _filter_aggregate?
+        # It's also probably overkill, since $out must be the last item in the pipeline.
+        pipeline = list(dictionary for dictionary in pipeline if "$out" not in dictionary) if pipeline else None
+        if 'filter.read' in app.config:
+            pipeline = _filter_aggregate(pipeline) if pipeline else None
+        pipeline = _convert_object(pipeline) if pipeline else None
+        limit = int(bottle.request.query.get('limit', 100))
+        pipeline = pipeline + [{'$limit': limit}] if pipeline else None
+        cursor = mongodb[collection].aggregate(pipeline=pipeline)
+        return {'results': [document for document in cursor]}
+
+
+@app.route('/<collection>', method=['GET'])
 def get(mongodb, collection):
-    filter_ = _get_json('filter')
-    projection = _get_json('projection')
-    skip = int(bottle.request.query.get('skip', 0))
-    limit = int(bottle.request.query.get('limit', 100))
-    sort = _get_json('sort')
-    # Turns a list of lists to a list of tuples.
-    # This is necessary because JSON has no concept of "tuple" but pymongo
-    # takes a list of tuples for the sort order.
-    sort = [tuple(field) for field in sort] if sort else None
-
-    cursor = mongodb[collection].find(
-        filter=filter_, projection=projection, skip=skip, limit=limit,
-        sort=sort
-    )
-
-    distinct = bottle.request.query.get('distinct')
-
-    if distinct:
-        return {'values': cursor.distinct(distinct)}
-
-    return {'results': [document for document in cursor]}
+    if bottle.request.method == 'GET':
+        filter_ = _get_json('filter')
+        projection = _get_json('projection')
+        skip = int(bottle.request.query.get('skip', 0))
+        limit = int(bottle.request.query.get('limit', 100))
+        sort = _get_json('sort')
+
+        # Turns a list of lists to a list of tuples.
+        # This is necessary because JSON has no concept of "tuple" but pymongo
+        # takes a list of tuples for the sort order.
+        sort = [tuple(field) for field in sort] if sort else None
+
+        filter_ = _convert_object(filter_) if filter_ else None
+
+        # We use a whitelist read setting to filter what is allowed to be read from the collection.
+        # If the whitelist read setting is empty or non existent, then nothing is filtered.
+        if os.environ.get('KALA_FILTER_READ') or 'filter.read' in app.config:
```

Comment:
This should be moved to application start up I think, it's wasted processing per-request since it will never change.